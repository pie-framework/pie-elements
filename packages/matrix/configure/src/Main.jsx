import React from 'react';
import PropTypes from 'prop-types';
import { EditableHtml } from '@pie-lib/pie-toolbox/editable-html';
import { InputContainer, settings, layout } from '@pie-lib/pie-toolbox/config-ui';
import { withStyles } from '@material-ui/core/styles';
import MatrixColumnsSizeHeaderInput from './MatrixColumnsSizeHeaderInput';
import MatrixRowsSizeHeaderInput from './MatrixRowsSizeHeaderInput';
import MatrixLabelTypeHeaderInput from './MatrixLabelTypeHeaderInput';
import MatrixValues from './MatrixValues';

const { Panel, toggle, radio } = settings;

const styles = (theme) => ({
  promptHolder: {
    width: '100%',
    paddingTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2,
  },
  matrixHeaderOptionsHolder: {
    display: 'flex',
    flexWrap: 'wrap',
    width: '100%',
    paddingBottom: theme.spacing.unit * 2.5,
    justifyContent: 'space-around',
  },
  errorText: {
    fontSize: theme.typography.fontSize - 2,
    color: theme.palette.error.main,
    paddingTop: theme.spacing.unit,
  },
});

const Design = withStyles(styles)((props) => {
  const {
    classes,
    model,
    configuration,
    onPromptChanged,
    imageSupport,
    uploadSoundSupport,
    onChangeModel,
    onConfigurationChanged,
    onTeacherInstructionsChanged,
  } = props;
  const {
    baseInputConfiguration = {},
    contentDimensions = {},
    prompt = {},
    scoringType = {},
    settingsPanelDisabled,
    spellCheck = {},
    teacherInstructions = {},
  } = configuration || {};
  const { errors = {}, extraCSSRules, teacherInstructionsEnabled, spellCheckEnabled } = model || {};
  const { prompt: promptError, teacherInstructions: teacherInstructionsError } = errors;

  const panelProperties = {
    teacherInstructionsEnabled: teacherInstructions.settings && toggle(teacherInstructions.label),
    spellCheckEnabled: spellCheck.settings && toggle(spellCheck.label),
    scoringType: scoringType.settings && radio(scoringType.label, ['auto', 'rubric']),
  };

  const getPluginProps = (props = {}) => ({
    ...baseInputConfiguration,
    ...props,
  });

  return (
    <layout.ConfigLayout
      extraCSSRules={extraCSSRules}
      dimensions={contentDimensions}
      hideSettings={settingsPanelDisabled}
      settings={
        <Panel
          model={model}
          onChangeModel={onChangeModel}
          configuration={configuration}
          onChangeConfiguration={onConfigurationChanged}
          groups={{
            Properties: panelProperties,
          }}
        />
      }
    >
      {teacherInstructionsEnabled && (
        <InputContainer label={teacherInstructions.label} className={classes.promptHolder}>
          <EditableHtml
            className={classes.prompt}
            markup={model.teacherInstructions || ''}
            onChange={onTeacherInstructionsChanged}
            imageSupport={imageSupport}
            error={teacherInstructionsError}
            pluginProps={getPluginProps(teacherInstructions?.inputConfiguration)}
            nonEmpty={false}
            spellCheck={spellCheckEnabled}
            uploadSoundSupport={uploadSoundSupport}
          />
          {teacherInstructionsError && <div className={classes.errorText}>{teacherInstructionsError}</div>}
        </InputContainer>
      )}

      <InputContainer label={prompt.label} className={classes.promptHolder}>
        <EditableHtml
          className={classes.prompt}
          markup={model.prompt}
          onChange={onPromptChanged}
          imageSupport={imageSupport}
          error={promptError}
          pluginProps={getPluginProps(prompt?.inputConfiguration)}
          nonEmpty={!prompt.settings}
          spellCheck={spellCheckEnabled}
          uploadSoundSupport={uploadSoundSupport}
          disableUnderline
        />
        {promptError && <div className={classes.errorText}>{promptError}</div>}
      </InputContainer>

      <div className={classes.matrixHeaderOptionsHolder}>
        <MatrixRowsSizeHeaderInput model={model} onChangeModel={onChangeModel} />
        <MatrixColumnsSizeHeaderInput model={model} onChangeModel={onChangeModel} />
        <MatrixLabelTypeHeaderInput model={model} onChangeModel={onChangeModel} />
      </div>

      <MatrixValues model={model} onChangeModel={onChangeModel} />
    </layout.ConfigLayout>
  );
});

export class Main extends React.Component {
  static propTypes = {
    model: PropTypes.object.isRequired,
    disableSidePanel: PropTypes.bool,
    onModelChanged: PropTypes.func.isRequired,
    onConfigurationChanged: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired,
    }),
  };

  onPromptChanged = (prompt) => {
    this.props.onModelChanged({ ...this.props.model, prompt });
  };

  onChangeModel = (data) => {
    this.props.onModelChanged({ ...this.props.model, ...data });
  };

  onTeacherInstructionsChanged = (teacherInstructions) => {
    this.props.onModelChanged({ ...this.props.model, teacherInstructions });
  };

  render() {
    return (
      <Design
        {...this.props}
        onChangeModel={this.props.onModelChanged}
        onPromptChanged={this.onPromptChanged}
        onTeacherInstructionsChanged={this.onTeacherInstructionsChanged}
      />
    );
  }
}

const Styled = withStyles(styles)(Main);

export default Styled;
