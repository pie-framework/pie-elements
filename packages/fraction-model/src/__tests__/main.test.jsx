import { shallow } from 'enzyme';
import React from 'react';
import Main from '../main';
import { shallowChild } from '@pie-lib/pie-toolbox/test-utils';
import AnswerFraction from '../answer-fraction';

describe('Main', () => {
  const defaultProps = {
    model: {
      correctResponse: [],
      title: '',
      question: '',
      modelTypeSelected: 'bar',
      maxModelSelected: 1,
      partsPerModel: 5,
      allowedStudentConfig: false,
      showGraphLabels: false,
    },
    session: {},
    onSessionChange: jest.fn(),
  };

  let wrapper;
  let component;

  beforeEach(() => {
    wrapper = shallowChild(Main, defaultProps, 1);
  });

  describe('render', () => {
    let w;

    beforeEach(() => {
      w = (props) => shallow(<Main {...props} />);
    });

    it('snapshot', () => {
      expect(w(defaultProps)).toMatchSnapshot();
    });

    it('renders correctly', () => {
      component = wrapper();

      expect(component.find(AnswerFraction).length).toEqual(1);

      expect(component.state()).toEqual({
        answerChangeDialog: {
          open: false,
          text: '',
        },
        fractionModelKey: 0,
        session: {
          answers: {
            response: [],
          },
        },
        showCorrect: false,
      });
    });

    it('renders correctly with a pre-filled session', () => {
      component = wrapper();
      expect(component.find(AnswerFraction).length).toEqual(1);
      component = wrapper({
        session: {
          answers: {
            noOfModel: 2,
            partsPerModel: 4,
            response: [
              { id: 1, value: 4 },
              { id: 2, value: 3 },
            ],
          },
        },
      });
      expect(component.find(AnswerFraction).length).toEqual(1);
      expect(component.state()).toEqual({
        answerChangeDialog: {
          open: false,
          text: '',
        },
        fractionModelKey: 0,
        session: {
          answers: {
            noOfModel: 2,
            partsPerModel: 4,
            response: [
              { id: 1, value: 4 },
              { id: 2, value: 3 },
            ],
          },
        },
        showCorrect: false,
      });
    });

    it('generate answers correctly from selection', () => {
      component = wrapper();
      expect(
        component.instance().generateAnswers({
          allowedStudentConfig: true,
        }),
      ).toEqual({
        noOfModel: 0,
        partsPerModel: 0,
        response: [],
      });
    });

    it('on answer change correctly after selection', () => {
      component = wrapper();
      component.instance().onAnswerChange({
        noOfModel: 4,
        partsPerModel: 2,
        response: [
          { id: 1, value: 2 },
          { id: 2, value: 2 },
          { id: 3, value: 2 },
          { id: 4, value: 1 },
        ],
      });
      expect(component.state()).toEqual({
        session: {
          answers: {
            response: [],
          },
        },
        showCorrect: false,
        answerChangeDialog: {
          newSession: {
            answers: {
              noOfModel: 4,
              partsPerModel: 2,
              response: [
                { id: 1, value: 2 },
                { id: 2, value: 2 },
                { id: 3, value: 2 },
                { id: 4, value: 1 },
              ],
            },
          },
          oldSession: {
            answers: {
              response: [],
            },
          },
          open: true,
          text: 'Changing either the Number of Models or Parts per Model will remove added answer. Are you sure you want to continue?',
        },
        fractionModelKey: 0,
      });
    });
  });
});
