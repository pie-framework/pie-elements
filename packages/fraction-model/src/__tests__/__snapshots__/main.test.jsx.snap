// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main render snapshot 1`] = `
<Main
  classes={
    Object {
      "modelPreview": "Main-modelPreview-2",
      "previewHeading": "Main-previewHeading-1",
      "titleContainer": "Main-titleContainer-3",
    }
  }
  model={
    Object {
      "allowedStudentConfig": false,
      "correctResponse": Array [],
      "maxModelSelected": 1,
      "modelTypeSelected": "bar",
      "partsPerModel": 5,
      "prompt": "",
      "showGraphLabels": false,
      "title": "",
    }
  }
  onSessionChange={[MockFunction]}
  session={Object {}}
/>
`;
